packages/core/src/style-collector.ts

```diff
@@ -0,0 +1,282 @@
+      this.atomic.add(styleResult)
+      this.classNames.set(styleResult.className, styleResult)
+    })
+
+    // no need to sort, each recipe is scoped using recipe.className
+    hashFactory.recipes.forEach((set, recipeName) => {
+      const recipeConfig = this.context.recipes.getConfig(recipeName)
+      if (!recipeConfig) return
+
+      set.forEach((item) => {
+        const process = (hash: string) => {
+          const styleResult = this.getAtomic(hash)
+          if (!styleResult) return
+
+          const stylesSet = getOrCreateSet(this.recipes, recipeName)
+          stylesSet.add(styleResult)
+
+          this.classNames.set(styleResult.className, styleResult)
+        }
+
+        if ('slots' in recipeConfig) {
+          recipeConfig.slots.forEach((slot) => process(item + HashFactory.separator + 'slot:' + slot))
+        } else {
+          process(item)
+        }
+      })
+    })
+
+    hashFactory.recipes_base.forEach((set, recipeKey) => {
+      const [recipeName, slot] = recipeKey.split(this.context.recipes.slotSeparator)
+      const recipeConfig = this.context.recipes.getConfig(recipeName)
+      if (!recipeConfig) return
+
+      const styleResult = this.getRecipeBase(set, recipeName, slot)
+      if (!styleResult) return
+
+      const stylesSet = getOrCreateSet(this.recipes_base, recipeKey)
+      stylesSet.add(styleResult)
+
+      this.classNames.set(styleResult.className, styleResult)
+    })
+
+    return this
+  }
+}
+
+const entryKeys = ['cond', 'recipe', 'layer', 'slot'] as const
+const getEntryFromHash = (hash: string) => {
+  const parts = hash.split(HashFactory.separator)
+  const prop = parts[0]
+
+  const rawValue = parts[1].replace('value:', '')
+  const parsed = Number(rawValue)
+  const value = parsed ? parsed : rawValue
+
+  const entry = { prop, value } as StyleEntry
+
+  parts.forEach((part) => {
+    const key = entryKeys.find((k) => part.startsWith(k))
+    if (key) {
+      entry[key] = part.slice(key.length + 1)
+    }
+  })
+
+  return entry
+}
+
+const makeObjAt = (path: string[], value: Record<string, unknown>) => {
+  if (!path.length) return value as StyleResultObject
+
+  const obj = {} as StyleResultObject
+  let current = obj
+  for (let i = 0; i < path.length; i++) {
+    const key = path[i]
+    if (i === path.length - 1) {
+      current[key] = value
+    } else {
+      current[key] = {}
+      current = current[key] as StyleResultObject
+    }
+  }
+
+  return obj
+}
+
+const setValueIn = (obj: StyleResultObject, path: string[], value: Record<string, unknown>) => {
+  if (!path.length) return Object.assign(obj, value) as StyleResultObject
+
+  let current = obj as Record<string, any>
+  for (let i = 0; i < path.length; i++) {
+    const key = path[i]
+    if (!current[key]) {
+      current[key] = {}
+    }
+
+    if (i === path.length - 1) {
+      Object.assign(current[key], value)
+    } else {
+      current = current[key]
+    }
+  }
+
+  return obj
+}
```

Comment:
Can we move these out of the file? Either to the shared package or to a different files